File: flutter/PulsarLottie/lib/src/haptic_lottie_controller.dart
import 'dart:async';

import 'package:flutter/widgets.dart';
import 'package:pulsar_haptics/pulsar.dart';

import 'sampler.dart';

/// How the haptics are produced while the animation plays.
enum HapticMode {
  /// The animation timeline is the master clock: the pattern is sampled every
  /// frame into `RealtimeComposer` events. Honours pause/seek/loop. Requires a
  /// [PatternData] source.
  realtime,

  /// A whole pattern is played once via `PatternComposer`, aligned to the start
  /// (best native fidelity). Seek/pause on the haptic side are best-effort.
  pattern,
}

/// Drives Pulsar haptics from a Lottie [AnimationController].
///
/// Attach it to the `AnimationController` you already pass to the `Lottie`
/// widget; transport ([play]/[pause]/[resume]/[stop]/[reset]/[setTimestamp]/
/// [setLoop]) steers both the animation and the haptics. In [HapticMode.realtime]
/// the controller listens to the animation and samples the pattern; in
/// [HapticMode.pattern] it fires a pre-parsed pattern aligned to the start.
///
/// The `AnimationController` is owned by the caller — this class does not
/// dispose it. Call [dispose] to detach and release the haptic resources.
class HapticLottieController {
  /// Creates a controller bound to [animationController].
  ///
  /// Pass a bundle [preset] to take its pattern and authored duration, or
  /// [haptics] for a pattern of your own — an explicit [haptics] wins. Omit both
  /// for a plain animation. A preset that carries audio plays through its own
  /// handle, which needs [HapticMode.pattern], so [hapticMode] defaults to
  /// `pattern` for one; pass it yourself to override. The [pulsar] instance is
  /// created internally if not supplied.
  HapticLottieController({
    required this.animationController,
    this.preset,
    this.haptics,
    HapticMode? hapticMode,
    this.hapticOffset = 0,
    this.hapticsEnabled = true,
    this.durationMs,
    Pulsar? pulsar,
  }) : _pulsar = pulsar ?? Pulsar(),
       _playsItself = haptics == null && (preset?.hasAudio ?? false),
       hapticMode =
           hapticMode ??
           (haptics == null && (preset?.hasAudio ?? false)
               ? HapticMode.pattern
               : HapticMode.realtime) {
    _attach();
  }

  /// The Lottie animation clock this controller follows and steers.
  final AnimationController animationController;

  /// Bundle preset supplying the pattern and the authored duration. `null` ⇒ none.
  final PresetHandle? preset;

  /// Pattern to sync with the animation. Overrides [preset]'s own pattern.
  final PatternData? haptics;

  /// Engine mode. Defaults to [HapticMode.realtime], or to [HapticMode.pattern]
  /// for a [preset] that carries audio.
  final HapticMode hapticMode;

  /// Device tuning: shift haptics by ±ms relative to the animation.
  final double hapticOffset;

  /// When `false`, the animation still plays but no haptics are emitted.
  final bool hapticsEnabled;

  /// Explicit clock length in ms. Overrides every derived duration.
  final double? durationMs;

  final Pulsar _pulsar;
  final bool _playsItself;
  PulsarRealtimeComposer? _realtime;
  PulsarPatternComposer? _pattern;
  double _lastT = 0;
  bool _disposed = false;

  /// The pattern actually driving the haptics: an explicit one, else the preset's.
  PatternData? get _resolvedPattern => haptics ?? preset?.pattern;

  bool get _useRealtime =>
      hapticMode == HapticMode.realtime && _resolvedPattern != null;

  bool get _hasContinuous {
    final pattern = _resolvedPattern;
    return pattern != null &&
        pattern.continuousPattern.amplitude.isNotEmpty &&
        pattern.continuousPattern.frequency.isNotEmpty;
  }

  /// Effective clock length in ms: an explicit [durationMs], else the preset's
  /// authored duration, else the Lottie composition once loaded, else the
  /// pattern's own length.
  double get durationMsResolved {
    final explicit = durationMs;
    if (explicit != null && explicit > 0) {
      return explicit;
    }
    final authored = preset?.duration;
    if (authored != null && authored > 0) {
      return authored;
    }
    final d = animationController.duration;
    if (d != null && d.inMicroseconds > 0) {
      return d.inMicroseconds / 1000.0;
    }
    final pattern = _resolvedPattern;
    return pattern != null ? patternDurationMs(pattern) : 0;
  }

  void _attach() {
    if (_resolvedPattern == null || !hapticsEnabled) {
      return;
    }
    if (_useRealtime) {
      _realtime = _pulsar.getRealtimeComposer();
      animationController.addListener(_onTick);
    } else if (!_playsItself) {
      _pattern = _pulsar.getPatternComposer();
      // Pre-parse so the engine is warm and play() fires without delay.
      unawaited(_pattern!.parsePattern(_resolvedPattern!));
    }
  }

  void _onTick() {
    final pattern = _resolvedPattern;
    if (_disposed || !hapticsEnabled || !_useRealtime || pattern == null) {
      return;
    }
    final dur = durationMsResolved;
    final t = animationController.value * dur;
    final ht = t + hapticOffset;
    if (_hasContinuous) {
      unawaited(
        _realtime!.set(
          clamp01(sampleEnvelope(pattern.continuousPattern.amplitude, ht)),
          clamp01(sampleEnvelope(pattern.continuousPattern.frequency, ht)),
        ),
      );
    }
    var prev = _lastT;
    if (t < prev) {
      prev = 0; // wrapped on loop
    }
    for (final e in pattern.discretePattern) {
      if (e.time > prev && e.time <= t) {
        unawaited(
          _realtime!.playDiscrete(clamp01(e.amplitude), clamp01(e.frequency)),
        );
      }
    }
    _lastT = t;
  }

  Future<void> _fireHaptics() async {
    if (!hapticsEnabled || _resolvedPattern == null) {
      return;
    }
    if (_useRealtime) {
      _lastT = 0;
    } else if (_playsItself) {
      preset!.play();
    } else {
      await _pattern?.play();
    }
  }

  Future<void> _stopHaptics() async {
    if (_useRealtime) {
      _lastT = 0;
      if (_hasContinuous) {
        await _realtime?.stop();
      }
    } else if (_playsItself) {
      preset?.stop();
    } else {
      await _pattern?.stop();
    }
  }

  /// Play from the start, animation and haptics together.
  Future<void> play() async {
    _lastT = 0;
    unawaited(animationController.forward(from: 0));
    await _fireHaptics();
  }

  /// Pause both animation and haptics.
  Future<void> pause() async {
    animationController.stop();
    await _stopHaptics();
  }

  /// Resume from the current position.
  Future<void> resume() async {
    unawaited(animationController.forward());
  }

  /// Stop and rewind to the start.
  Future<void> stop() async {
    animationController.reset();
    _lastT = 0;
    await _stopHaptics();
  }

  /// Rewind to the start (also stops haptics).
  Future<void> reset() async {
    animationController.reset();
    _lastT = 0;
    await _stopHaptics();
  }

  /// Seek both animation and haptics to [ms] from the start.
  void setTimestamp(double ms) {
    final dur = durationMsResolved;
    if (dur > 0) {
      animationController.value = clamp01(ms / dur);
    }
    _lastT = ms;
  }

  /// Loop the animation. [count] limits the iterations (null = forever);
  /// [reverse] plays a boomerang. In `pattern` mode the pattern fires once at
  /// the start (per-iteration re-fire is not automatic).
  Future<void> setLoop(bool loop, {int? count, bool reverse = false}) async {
    if (loop) {
      _lastT = 0;
      unawaited(animationController.repeat(reverse: reverse, count: count));
      await _fireHaptics();
    } else {
      animationController.stop();
    }
  }

  /// Detach from the animation and release haptic resources. Does **not**
  /// dispose [animationController] — the caller owns it.
  void dispose() {
    if (_disposed) {
      return;
    }
    _disposed = true;
    if (_useRealtime) {
      animationController.removeListener(_onTick);
    }
    unawaited(_stopHaptics());
    unawaited(_pattern?.dispose());
  }
}

File: Android/PulsarLottie/src/main/java/com/swmansion/pulsar/lottie/HapticLottieController.kt
package com.swmansion.pulsar.lottie

import android.animation.ValueAnimator
import com.airbnb.lottie.LottieAnimationView
import com.airbnb.lottie.LottieDrawable
import com.swmansion.pulsar.Pulsar
import com.swmansion.pulsar.bundle.PresetHandle
import com.swmansion.pulsar.composers.PatternComposer
import com.swmansion.pulsar.composers.RealtimeComposer
import com.swmansion.pulsar.types.PatternData

/** How the haptics are produced while the animation plays. */
enum class HapticMode {
    /**
     * The animation timeline is the master clock: the pattern is sampled every
     * frame into `RealtimeComposer` events. Honours pause/seek/loop. Requires a
     * [PatternData] source.
     */
    REALTIME,

    /**
     * A whole pattern is played once via `PatternComposer`, aligned to the start
     * (best native fidelity). Seek/pause on the haptic side are best-effort.
     */
    PATTERN,
}

/**
 * Drives Pulsar haptics from a [LottieAnimationView].
 *
 * Attach it to a `LottieAnimationView` you already use; the transport
 * ([play]/[pause]/[resume]/[stop]/[reset]/[setTimestamp]/[setLoop]) steers both
 * the animation and the haptics. In [HapticMode.REALTIME] it follows the view's
 * animator (the per-frame clock) and samples the pattern; in [HapticMode.PATTERN]
 * it fires a pre-parsed pattern aligned to the start.
 *
 * A bundle [preset] supplies the pattern and authored duration; an explicit [haptics]
 * overrides it. [hapticMode] defaults to [HapticMode.REALTIME], or to [HapticMode.PATTERN]
 * for a preset carrying audio, which only sounds there.
 *
 * Call [release] when done to detach the animator listener and stop haptics.
 */
class HapticLottieController @JvmOverloads constructor(
    private val lottieView: LottieAnimationView,
    pulsar: Pulsar,
    preset: PresetHandle? = null,
    haptics: PatternData? = null,
    hapticMode: HapticMode? = null,
    private val hapticOffset: Long = 0L,
    private val hapticsEnabled: Boolean = true,
    durationMs: Long? = null,
) {
    private val pattern: PatternData? = haptics ?: preset?.pattern
    private val playsOwnAudio = haptics == null && preset?.hasAudio == true
    private val mode = hapticMode ?: if (playsOwnAudio) HapticMode.PATTERN else HapticMode.REALTIME

    private val useRealtime = mode == HapticMode.REALTIME && pattern != null
    private val hasContinuous = pattern != null &&
        pattern.continuousPattern.amplitude.isNotEmpty() &&
        pattern.continuousPattern.frequency.isNotEmpty()

    private val realtime: RealtimeComposer? =
        if (useRealtime) pulsar.getRealtimeComposer() else null

    private val audioPreset: PresetHandle? = if (!useRealtime && playsOwnAudio) preset else null

    private val composer: PatternComposer? =
        if (!useRealtime && !playsOwnAudio && pattern != null) {
            // Pre-parse so the engine is warm and play() fires without delay.
            pulsar.getPatternComposer().apply { parsePattern(pattern) }
        } else {
            null
        }

    /**
     * Clock length in ms: an explicit duration, else the preset's authored one, else the
     * Lottie composition once loaded, else the pattern's own length.
     */
    private val fixedDurationMs: Long? =
        durationMs?.takeIf { it > 0L } ?: preset?.duration?.takeIf { it > 0L }
    private var resolvedDurationMs: Long = fixedDurationMs
        ?: pattern?.let { patternDurationMs(it) }
        ?: 0L
    private var lastT: Long = 0L

    private val updateListener = ValueAnimator.AnimatorUpdateListener { anim ->
        onTick(anim.animatedFraction)
    }

    init {
        if (pattern != null && hapticsEnabled) {
            if (fixedDurationMs == null) {
                lottieView.addLottieOnCompositionLoadedListener { composition ->
                    resolvedDurationMs = composition.duration.toLong()
                }
            }
            if (useRealtime) lottieView.addAnimatorUpdateListener(updateListener)
        }
    }

    private fun onTick(fraction: Float) {
        if (!hapticsEnabled || !useRealtime || pattern == null) return
        val t = (fraction * resolvedDurationMs).toLong()
        val ht = t + hapticOffset
        if (hasContinuous) {
            realtime?.set(
                clamp01(sampleEnvelope(pattern.continuousPattern.amplitude, ht)),
                clamp01(sampleEnvelope(pattern.continuousPattern.frequency, ht)),
            )
        }
        var prev = lastT
        if (t < prev) prev = 0L // wrapped on loop
        for (e in pattern.discretePattern) {
            if (e.time > prev && e.time <= t) {
                realtime?.playDiscrete(clamp01(e.amplitude), clamp01(e.frequency))
            }
        }
        lastT = t
    }

    private fun fireHaptics() {
        if (!hapticsEnabled) return
        when {
            useRealtime -> lastT = 0L
            audioPreset != null -> audioPreset.play()
            else -> composer?.play()
        }
    }

    private fun stopHaptics() {
        when {
            useRealtime -> {
                lastT = 0L
                if (hasContinuous) realtime?.stop()
            }
            audioPreset != null -> audioPreset.stop()
            else -> composer?.stop()
        }
    }

    /** Play from the start, animation and haptics together. */
    fun play() {
        lastT = 0L
        lottieView.progress = 0f
        lottieView.playAnimation()
        fireHaptics()
    }

    /** Pause both animation and haptics. */
    fun pause() {
        lottieView.pauseAnimation()
        stopHaptics()
    }

    /** Resume from the current position. */
    fun resume() {
        lottieView.resumeAnimation()
    }

    /** Stop and rewind to the start. */
    fun stop() {
        lottieView.cancelAnimation()
        lottieView.progress = 0f
        lastT = 0L
        stopHaptics()
    }

    /** Rewind to the start (also stops haptics). */
    fun reset() = stop()

    /** Seek both animation and haptics to [ms] from the start. */
    fun setTimestamp(ms: Long) {
        if (resolvedDurationMs > 0L) {
            lottieView.progress = clamp01(ms.toFloat() / resolvedDurationMs)
        }
        lastT = ms
    }

    /**
     * Loop the animation. [count] limits iterations ([LottieDrawable.INFINITE] =
     * forever); [reverse] plays a boomerang.
     */
    @JvmOverloads
    fun setLoop(loop: Boolean, count: Int = LottieDrawable.INFINITE, reverse: Boolean = false) {
        lottieView.repeatCount = if (loop) count else 0
        lottieView.repeatMode = if (reverse) LottieDrawable.REVERSE else LottieDrawable.RESTART
    }

    /** Detach the animator listener and release haptic resources. */
    fun release() {
        if (useRealtime) lottieView.removeUpdateListener(updateListener)
        stopHaptics()
    }
}

/**
 * Attach Pulsar haptics to this [LottieAnimationView] without swapping the view.
 * Returns a [HapticLottieController] you drive; call [HapticLottieController.release]
 * when done.
 *
 * This binds haptics only — the view keeps whatever animation you gave it. Use
 * [HapticLottieView.bindHaptics] to have a bundle [preset]'s Lottie rendered for you.
 */
@JvmOverloads
fun LottieAnimationView.bindHaptics(
    pulsar: Pulsar,
    preset: PresetHandle? = null,
    haptics: PatternData? = null,
    hapticMode: HapticMode? = null,
    hapticOffset: Long = 0L,
    hapticsEnabled: Boolean = true,
    durationMs: Long? = null,
): HapticLottieController = HapticLottieController(
    this,
    pulsar,
    preset,
    haptics,
    hapticMode,
    hapticOffset,
    hapticsEnabled,
    durationMs,
)
